### PR TITLE
Gen2 - Fix detached thread issues

### PR DESCRIPTION
**Pull request by themarpe**
EDITED since original PR
 - This removes thread detaching, doesn't use timeout calls. 
 - Uses XLinkResetRemote to gracefully shutdown XLink connection. 
 - Moved XLinkStream operations to its own class
 - Fixed timeout `send` in `DataInputQueue` not throwing if data size is too large
 - Added `close` and `isClosed` to `Device` and `DeviceBootloader`
 - Modified shutdown, instead of static sleep, `XLinkConnection` wait until the device comes back online (with 5s timeout)

### CODE REVIEW

**Review by szabi-luxonis — APPROVED** [review]
LGTM

**Comment by szabi-luxonis** [comment]
Pleae open a PR on xlink repo too, for cross linking.

File: src/device/DataQueue.cpp [hunk 1]
```diff
@@ -170,79 +155,63 @@ bool DataOutputQueue::removeCallback(int callbackId) {
...
-                    if(!metadata.empty()) objData = nlohmann::json::from_msgpack(metadata).dump();
-                    spdlog::trace("Sending message to device - data size: {}, object type: {} object data: {}", data->data.size(), type, objData);
-                }
+DataInputQueue::DataInputQueue(const std::shared_ptr<XLinkConnection>& conn, const std::string& streamName, unsigned int maxSize, bool blocking)
+    : queue(maxSize, blocking), name(streamName) {
+    // open stream with default XLINK_USB_BUFFER_MAX_SIZE write size
+    XLinkStream stream(*conn, name, dai::XLINK_USB_BUFFER_MAX_SIZE);
+
+    writingThread = std::thread(std::bind(
+        [this, conn](XLinkStream& stream) {
```
Comment (szabi-luxonis) [inline src/device/DataQueue.cpp:164]: why writingThread is not mutable, but readingThread is ?

File: src/device/Device.cpp [hunk 2]
```diff
@@ -386,41 +426,41 @@ void Device::init(const Pipeline& pipeline, bool embeddedMvcmd, bool usb2Mode, c
...
-        } catch(const std::runtime_error& err) {
-            // ignore
-            spdlog::debug("Watchdog thread exception caught: {}", err.what());
-        }
+        // Watchdog ended. Useful for checking disconnects
+        watchdogRunning = false;
 
-        // Sleep a bit, so device isn't available anymore
-        std::this_thread::sleep_for(std::chrono::milliseconds(500));
+        // // reset device
```
Comment (szabi-luxonis) [inline src/device/Device.cpp:432]: remove commented out code

File: src/device/DataQueue.cpp (hunk 1 above)
Comment (themarpe) [reply src/device/DataQueue.cpp:164]: Forgot to remove after some tests. Fixed.

**Review by szabi-luxonis — APPROVED** [review]

**Pull request merged**